Delete Dog API Endpoint

parkerra opened the pull request:
## Delete Dog Endpoint
Issue Number: #7 

The API endpoint DELETE dog/[:id] has been created which verifies if the ID is a valid MongoDB ObjectID, and then calls the database action deleteDog. This action deletes the dog if the ID is present in the database, returns null if the ID does not exist, or throws an error if there is an error with the database.

### Checklist
- [x] Requirements have been implemented
- [x] Acceptance criteria is met
- [x] Database schema docs have been updated or are not necessary
- [x] Code follows design and style guidelines
- [x] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (EM) have been assigned to this PR

### How To Test
Use Postman with a DELETE request after starting server using `npm run dev`. Then try different cases with the dog's ObjectId in the query parameters in the URL using http://localhost:3000/api/dogs/[:id]

Cases to Test:
* Successfully deleting dog with a valid ObjectId
* Unsuccessfully deleting dog with valid ObjectId not present in database
* Unsuccessfully deleting dog with invalid ObjectId (such as `0` or `abcdefg`)

parkerra committed:
Add delete dog endpoint and database actions

The API endpoint DELETE dog/[:id] has been created and calls
the database action deleteDog. This actions checks if the ID
is valid and deletes the dog if it is, returns null if the
ID does not exist, or throws an error if there is an error
with the database.

afazio1 commented on src/pages/api/dogs/[id].js:
```diff
@@ -0,0 +1,26 @@
+      .then((results) => {
+        if (!results) {
+          return res
+            .status(422)
```
I think `404 - Resource Not Found` status would make more sense here

afazio1 commented on src/pages/api/dogs/[id].js:
```diff
@@ -0,0 +1,26 @@
+    if (!mongoose.isValidObjectId(req.query.id)) {
+      return res
+        .status(422)
+        .send("Unable to delete because dog id is not in valid format");
```
Let's have the api return json in the form:
1. Error occurs
```js
{
  success: false,
  message: "Unable to delete because dog id is not in valid format",
}
```
2. Successful res with data
```js
{
   success: true,
   message: "Dog successfully created",
   data: dogId
}
```

afazio1 reviewed (changes requested):
To do: Change one status code and clarify what api endpoints should be returning (json shape).
Error handling is very thorough, love the js doc comments, great work 🤩

parkerra committed:
Change API responses to JSON

The API responses were previously strings so they have been
converted to JSON objects specifying the success state,
message, and data if applicable.

afazio1 commented on src/pages/api/dogs/[id].js:
```diff
@@ -0,0 +1,35 @@
+          return res.status(200).send({
+            success: true,
+            message: "Dog sucessfully deleted",
+            data: results,
```
Let's just return the `{ _id }` of the dog upon success. I don't think our frontend will need the whole dog object

afazio1 reviewed (changes requested):
Two minor changes with the data returned on success and ending the request if an invalid method is used. After that this looks great to merge!

afazio1 commented on src/pages/api/dogs/[id].js:
```diff
@@ -0,0 +1,35 @@
+          message: "Unable to delete dog, please try again",
+        });
+      });
+  }
```
We should also end the request if they do not use an "allowed" method. Outside the `if (req.method == "DELETE")` statement, you can add:

```js
res.status(405).send({success: false, message: `Request method ${req.method} is not allowed`})
```

parkerra committed:
Return deleted dog id instead of whole object

Instead of returning the whole dog object by the API, just the id of
the deleted dog is returned. Also, connecting to the database has been
moved inside a try-catch block so that any errors with connecting to
the database can be caught.

afazio1 committed:
Merge branch 'dev' into parker-7-delete_dog_endpoint
